lib/remap-lang/src/type_def.rs

```diff
@@ -16,13 +16,42 @@ pub struct TypeDef {
 
     /// The [`value::Kind`]s this definition represents.
     pub kind: value::Kind,
+
+    /// Some types contain a collection of other types. If they do, this value
+    /// is set to `Some`, and returns the [`TypeDef`] of the collected inner
+    /// types.
+    ///
+    /// For example, given a [`Value::Array`]:
+    ///
+    /// ```rust
+    /// # use remap_lang::{expression::Array, Value, Expression, state, TypeDef, value::Kind};
+    ///
+    /// let vec = vec![Value::Null, Value::Boolean(true)];
+    /// let expression = Array::from(vec);
+    /// let state = state::Compiler::default();
+    ///
+    /// assert_eq!(
+    ///     expression.type_def(&state),
+    ///     TypeDef {
+    ///         fallible: false,
+    ///         kind: Kind::Array,
+    ///         inner_type_def: Some(TypeDef {
+    ///             fallible: false,
+    ///             kind: Kind::Null | Kind::Boolean,
+    ///             inner_type_def: None,
+    ///         }.boxed()),
+    ///     },
+    /// );
+    /// ```
+    pub inner_type_def: Option<Box<TypeDef>>,
```

Comment:
One thing I should note here is that we don't really have a reason to store `TypeDef`, we could just store `Kind`, but then we'd have to move (and rename) `inner_type_def` to add it to `Kind` (e.g. `kind.inner_kind`), but that in term is problematic because `Kind` is currently just an integer (using bitflags).

I started making this change, but the churn was too great to justify for a first iteration.

Basically, the result of this is that we get `fallible` on every level, even though there's no way to have an inner type be fallible, but an outer type not to be fallible.

Perhaps we'll tidy this up in the near future, but I found it to be too small of an issue to solve in this PR.